be/src/storage/lake/spill_mem_table_sink.cpp

```diff
@@ -139,11 +147,102 @@ Status SpillMemTableSink::flush_chunk_with_deletes(const Chunk& upserts, const C
+        if (_block_idx >= _blocks.size()) {
+            return Status::EndOfFile("End of block group");
+        }
+        RETURN_IF_ERROR(_sink.read_chunk_from_group_block(chunk, *_blocks[_block_idx]));
+        _block_idx++;
+        return Status::OK();
+    }
+
+    void close() {}
+
+private:
+    SpillMemTableSink& _sink;
+    const std::vector<spill::BlockPtr>& _blocks;
+    size_t _block_idx = 0;
+};
+
+Status SpillMemTableSink::read_chunk_from_group_block(Chunk* chunk, spill::Block& block) {
+    spill::SerdeContext ctx;
+    spill::BlockReaderOptions options;
+    auto& metrics = _spiller->serde()->parent()->metrics();
+    options.read_io_timer = metrics.read_io_timer;
+    options.read_io_count = metrics.read_io_count;
+    options.read_io_bytes = metrics.restore_bytes;
+    auto reader = block.get_reader(options);
+    DCHECK(reader);
+    ASSIGN_OR_RETURN(auto tmp_chunk, _spiller->serde()->deserialize(ctx, reader.get()));
+    tmp_chunk->swap_chunk(*chunk);
+    return Status::OK();
+}
+
 Status SpillMemTableSink::merge_blocks_to_segments() {
-    // TODO
-    // 1. get blocks from `_block_manager`(LoadSpillBlockManager)
-    // 2. merge blocks to segments
-    // 3. add these segments to `_writer`(TabletWriter)
+    MonotonicStopWatch timer;
+    timer.start();
+    // merge process needs to control _writer's flush behavior manually
+    _writer->set_auto_flush(false);
+    auto char_field_indexes = ChunkHelper::get_char_field_indexes(*_schema);
+    size_t total_blocks = 0;
+    size_t total_block_bytes = 0;
+    size_t total_merges = 0;
+    size_t total_rows = 0;
+    size_t total_chunk = 0;
+
+    auto groups = _block_manager->block_container()->block_groups();
+    std::vector<ChunkIteratorPtr> merge_inputs;
+    size_t current_input_bytes = 0;
+    auto merge_func = [&] {
+        total_merges++;
+        auto merge_itr = new_heap_merge_iterator(merge_inputs);
+        RETURN_IF_ERROR(merge_itr->init_encoded_schema(EMPTY_GLOBAL_DICTMAPS));
+        auto chunk_shared_ptr = ChunkHelper::new_chunk(*_schema, config::vector_chunk_size);
+        auto chunk = chunk_shared_ptr.get();
+        while (true) {
+            chunk->reset();
+            auto st = merge_itr->get_next(chunk);
+            if (st.is_end_of_file()) {
+                break;
+            } else if (st.ok()) {
+                ChunkHelper::padding_char_columns(char_field_indexes, *_schema, _writer->tablet_schema(), chunk);
+                total_rows += chunk->num_rows();
+                total_chunk++;
+                RETURN_IF_ERROR(_writer->write(*chunk, nullptr));
+            } else {
+                return st;
+            }
+        }
+        merge_itr->close();
+        return _writer->flush();
+    };
+    for (size_t i = 0; i < groups.size(); ++i) {
+        auto& group = groups[i];
+        if (merge_inputs.size() > 0 && current_input_bytes + group.data_size() > config::load_spill_max_merge_bytes) {
```

Comment:
`group.data_size()` will be smaller than chunk memory size.

Reply:
config::load_spill_max_chunk_bytes * merge_inputs.size() <= config::load_spill_max_merge_bytes

Reply:
change in future PR